### PR TITLE
Log an error when a Task returns false but didn't log an error

### LINKED ISSUE #2036: Log our own error if a task returned false but didn't log any errors
If a task violates the implicit expectation that it should return false iff it logged any errors MSBuild fails the build, but doesn't indicate why.

We should consider at least adding a warning, but ideally failing the build (perhaps in strict mode?)

### PR DESCRIPTION
**Pull request by BenVillalobos**
Fixes #2036 

In the off chance that a task returns false but does not log an error, the build will _look_ like it succeeded despite a failure. This change raises an error displaying the task that returned false but did not log an error.

### CODE REVIEW

File: src/Build.UnitTests/BackEnd/BuildResult_Tests.cs [hunk 1]
```diff
@@ -10,19 +10,22 @@
 using Microsoft.Build.Framework;
 using Microsoft.Build.Shared;
 using Microsoft.Build.Unittest;
+using Shouldly;
 using Xunit;
-
+using Xunit.Sdk;
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]:
```suggestion
using Xunit.Abstractions;
```

File: src/Build.UnitTests/BackEnd/BuildResult_Tests.cs [hunk 2]
```diff
@@ -10,19 +10,22 @@
...
 using TaskItem = Microsoft.Build.Execution.ProjectItemInstance.TaskItem;
 
 namespace Microsoft.Build.UnitTests.BackEnd
 {
     public class BuildResult_Tests
     {
         private int _nodeRequestId;
+        TestOutputHelper _output;
 
         public BuildResult_Tests()
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]:
```suggestion
        public BuildResult_Tests(ITestOutputHelper output)
```

File: src/Build.UnitTests/BackEnd/BuildResult_Tests.cs [hunk 3]
```diff
@@ -10,19 +10,22 @@
...
         {
             _nodeRequestId = 1;
+            _output = new TestOutputHelper();
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]:
```suggestion
            _output = output;
```

File: src/Build.UnitTests/BackEnd/BuildResult_Tests.cs [hunk 4]
```diff
@@ -347,6 +350,27 @@ public void TestTranslation()
             Assert.True(TranslationHelpers.CompareCollections(result["omega"].Items, deserializedResult["omega"].Items, TaskItemComparer.Instance));
         }
 
+        [Fact]
+        public void TaskReturnsFailureButDoesNotLogErrorShouldCauseBuildFailure()
+        {
+
+            using (TestEnvironment env = TestEnvironment.Create())
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]:
```suggestion
            using (TestEnvironment env = TestEnvironment.Create(_output))
```

this doesn't usually help much but doesn't hurt either

File: src/Build.UnitTests/BackEnd/ReturnFailureWithoutLoggingErrorTask.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+namespace Microsoft.Build.UnitTests
+{
+    /// <summary>
+    /// This task was created for https://github.com/microsoft/msbuild/issues/2036
+    /// </summary>
+    public class ReturnFailureWithoutLoggingErrorTask : Task
+    {
+        public override bool Execute()
+        {
+            return false;
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/ReturnFailureWithoutLoggingErrorTask.cs]: Probably worth a comment explicitly mentioning not logging any error.

File: src/Build.UnitTests/BackEnd/ReturnFailureWithoutLoggingErrorTask.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿using Microsoft.Build.BackEnd.Logging;
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/ReturnFailureWithoutLoggingErrorTask.cs]: copyright header please

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs [hunk 7]
```diff
@@ -937,6 +940,14 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
                     }
                 }
 
+                // A task that returned false should have logged an error, otherwise log that as an error
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs]: extend comment to describe the msbuild-task thing

File: src/Build/Resources/Strings.resx [hunk 8]
```diff
@@ -1063,6 +1063,10 @@
       LOCALIZATION: "{2}" is a localized message from a CLR/FX exception. Also, Microsoft.Build.Framework should not be localized
     </comment>
   </data>
+  <data name="TaskReturnedFalseButDidNotLogError">
+    <value>MSB4132: Build failed because the "{0}" task returned false but did not log an error.</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx]:
```suggestion
    <value>MSB4132: The "{0}" task returned false but did not log an error.</value>
```

(task failures don't _always_ fail the build in the presence of `ContinueOnError`.

**Comment by BenVillalobos** [comment]
A lead on what's causing tests to hang:

```
..\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19517.3\tools\XUnit\XUnit.targets(89,5): warning MS
B5021: Terminating the task executable "cmd" and its child processes because the build was canceled. [..\Docu
ments\GitHub\msbuild\src\Build.UnitTests\Microsoft.Build.Engine.UnitTests.csproj]
..\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19517.3\tools\XUnit\XUnit.targets(89,5): error MSB4
132: Build failed because the "Exec" task returned false but did not log an error. [..\Documents\GitHub\msbui
ld\src\Build.UnitTests\Microsoft.Build.Engine.UnitTests.csproj]
```

**Comment by BenVillalobos** [comment]
The tests pass? I don't trust this, rerunning.

File: src/Build.UnitTests/BackEnd/BuildResult_Tests.cs (hunk 1 above)
Comment (BenVillalobos) [reply src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]: TestOutputHelper is defined in Xunit.Sdk

Comment (BenVillalobos) [reply src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]: Got it, switched to using ITestOutputHelper

File: src/Build/BackEnd/Components/Logging/LoggingContext.cs [hunk 9]
```diff
@@ -106,6 +109,8 @@ protected set
             }
         }
 
+        internal bool HasLoggedErrors { get { return _hasLoggedErrors; } set { _hasLoggedErrors = value; } }
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/Logging/LoggingContext.cs]: Should we add some more scoped unit tests to `LoggingContext_Tests`? It doesn't look like there are any tests that cover the methods you're changing :(

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs [hunk 10]
```diff
@@ -937,6 +941,16 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
                     }
                 }
 
+                // When a task fails it must log an error. If a task fails to do so,
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs]:
```suggestion
                // When a task fails it should have logged an error. If a task fails to do so,
```

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs [hunk 11]
```diff
@@ -937,6 +941,16 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
...
+                // that is logged as an error. MSBuild tasks are an exception because
+                // errors are not logged directly from them, but the tasks spawned by them.
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs]:
```suggestion
                // errors are not logged directly from them, but by tasks in the projects spawned by them.
```

File: src/Build.UnitTests/BackEnd/BuildResult_Tests.cs [hunk 12]
```diff
@@ -10,19 +10,22 @@
...
 
 namespace Microsoft.Build.UnitTests.BackEnd
 {
     public class BuildResult_Tests
     {
         private int _nodeRequestId;
+        ITestOutputHelper _output;
 
-        public BuildResult_Tests()
+        public BuildResult_Tests(ITestOutputHelper output)
```
[the remark this answers is not in the document]
Comment (BenVillalobos) [reply src/Build.UnitTests/BackEnd/BuildResult_Tests.cs]: No, this was more of a "I touched this file so might as well add this" change, despite undoing the need for the change now. I can remove it entirely at this point.

File: src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs [hunk 13]
```diff
@@ -262,5 +270,109 @@ private string GetTestProject(bool? treatAllWarningsAsErrors = null, string warn
...
+                    <UsingTask TaskName = ""ReturnFailureWithoutLoggingErrorTask"" AssemblyName=""Microsoft.Build.Engine.UnitTests""/>
+                    <Target Name='Build'>
+                        <ReturnFailureWithoutLoggingErrorTask
+                            ContinueOnError=""WarnAndContinue""/>
+                    </Target>
+                </Project>");
+
+                MockLogger logger = proj.BuildProjectExpectSuccess();
+
+                logger.AssertLogContains("MSB4132");
```
[the remark this answers is not in the document]
Comment (BenVillalobos) [reply src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs]: Interestingly enough: MockLogger doesn't see any warnings. It does however see that the build succeeded and that there is an error. So I'll instead check that:
1. the successfully project built.
1. The log contains the specific error I'm looking for
1. The error count is 1.

File: src/Build/BackEnd/Components/Logging/LoggingContext.cs (hunk 9 above)
Comment (BenVillalobos) [reply src/Build/BackEnd/Components/Logging/LoggingContext.cs]: Agreed. Any suggestions for testing this specifically?

File: src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs]: Does this work through real loggers, though?

[the remark this answers is not in the document]
Comment (BenVillalobos) [reply src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs]: PR Review Meeting: The expected behavior should be:
1. Project built
1. Error count should be 0
1. Log should contain msb4132

**Comment by BenVillalobos** [comment]
Testing azure pipeline syntax.

/azp run

File: src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (BenVillalobos) [reply src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs]: Fixed.

**Comment by BenVillalobos** [comment]
I'm okay with that. Getting a build to run and having access to the taskloggingcontext to see if it logged any errors was somewhat troublesome.

**Pull request merged**